### ISSUE TITLE
128bit hashing

### ISSUE THREAD

**Issue by RagnarGrootKoerkamp** [RagnarGrootKoerkamp/PTRHash#4]
Since this keeps popping up here's a new issue to track and discuss this.

Currently I only support 64bit hashes for simplicity and speed, but it seems that this will really not be sufficient for many/most usecases, so supporting 128bit hashes is important.

This could be in addition to existing 64bit hashes, or if the speed difference is minimal we could replace them completely.

Current situation (64bit hash):
- high bits of hash determine `part`
- next highest bits determine `bucket`
- the entire hash is used for the `slot`

PTHash (128bit hash):
- high 64bits determine `bucket`
- low 64bits determine `slot`

So the natural way to go would be:
- high bits of high 64-bit word determine `part`
- next highest bits of high 64-bit word determine `bucket`
- the low 64-bit word is used for the `slot`.

I think this doesn't fully exploit the 128 bits (I think would require mixing all 128 bits into the `slot`), but is mostly likely good enough (since PTHash already does this and it works).

**Comment by RagnarGrootKoerkamp** [comment]
@vigna @rob-p, do you have any specific 128bit hash function in mind? This is now basically supported but I have to find a good default hash function; ideally some rust crate that simply provides a simple function `hash(&[u8]) -> u128`.

**Comment by RagnarGrootKoerkamp** [comment]
Candidates:
- [Metrohash](https://docs.rs/metrohash/1.0.6/metrohash/struct.MetroHash128.html): supports `u64` seed by default
- [farmhash](https://docs.rs/farmhash/1.1.5/farmhash/) no `u128` output here :( 
- [SpookyHash](https://crates.io/crates/hashers) and others in `hashers` crate
- [wikipedia list](https://en.wikipedia.org/wiki/List_of_hash_functions)

**Comment by rob-p** [comment]
I generally like [ahash](https://docs.rs/ahash/latest/ahash/).

**Comment by RagnarGrootKoerkamp** [comment]
As far as I can see `ahash` implements the standard `Hasher` trait, but that only is [only able to return](https://doc.rust-lang.org/std/hash/trait.Hasher.html#tymethod.finish) 64bit hashes it seems.

**Comment by RagnarGrootKoerkamp** [comment]
Ok I implemented all of them in the [`hashers` branch](https://github.com/RagnarGrootKoerkamp/PTRHash/blob/hashers/src/hash.rs) and ran some quick experiments for sequential querying and querying with prefetching **on `u64` keys** (where inlining and compiler optimizations do matter a lot).

```
| hasher      | bits |  seq. | prefetch |
| fxhash      |   64 |  14.2 |     7.48 |
| murmur2     |   64 |  15.8 |     7.70 |
| fastmurmur3 |  128 |  16.5 |     9.40 |
| murmur3     |  128 |  84.5 |    27.50 |
| city        |   64 |  28.0 |    12.19 |
| city        |  128 |  45.1 |    20.09 |
| xx          |   64 |  39.0 |    14.04 |
| xx          |  128 |  32.3 |    13.94 |
| metro       |   64 |  29.9 |    13.21 |
| metro       |  128 |  40.3 |    16.71 |
| highway     |   64 | 107.6 |    57.20 |
| highway     |  128 | 114.5 |    47.84 |
| spooky      |   64 | 115.7 |    63.43 |
| spooky      |  128 | 115.5 |    67.06 |

```

I have no idea how efficient all specific implementations are, but some notes:
- `FxHash` (which is *very* simple) is the lower bound, and is basically memory bound.
- `Murmur2` (64bit) is only marginally slower.
- `fastmurmur3` (128bit) is slightly slower, but still good when needed.
- `murmur3` is waaay slower.
- `city`, `xx`, and `metro` all give ~2x slower performance than the fastest 3 options.
- `highway` and `spooky` are very slow

So takeaway:
- For 64bits use either `fxhash` or `murmur2`
- For 128bits use `fastmurmur3`.

I'll keep all others in the branch but not add them to master.

**Comment by vigna** [comment]
Maybe testing on strings, too? Like, urls?

**Comment by vigna** [comment]
Did you have a look here? https://github.com/rurban/smhasher

**Comment by vigna** [comment]
BTW, I didn't notice there was a 128-bit version of xxh3—I ditched SpookyHash in favor of that, with measurable advantages. It seems to be an excellent candidate for 128-bit output. WyHash is currently the fastest (among the good ones) for 64-bit output.

**Comment by RagnarGrootKoerkamp** [comment]
I added support for string keys now. Will push some benchmarks on hashes in a bit. One small issue though that I can't figure out:

Epserde doesn't like the `_key: PhantomData<&'k Key>` field that I had to add. Can you implement the required traits in epserde? Or should I put additional bounds on my `Key` type?

https://github.com/RagnarGrootKoerkamp/PTRHash/commit/1bafeeaffa8a742367c637ffe028a04b62150383#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R174

**Comment by vigna** [comment]
OMG, we just had this discussion yesterday—would anyone ever need to put a reference in a PhantomKey? It appears it might happen. I guess the issue is that you need the lifetime, there, too, right?

**Comment by vigna** [comment]
I'm puzzled as to why you need a reference there. The reference is needed only when you pass data in—for example, in VFunc type input type is str and then the iterator on input values returns &str. That is, the parameter for the keys should be the referenced type, not a reference. But maybe you need it that way for some other reason...

**Comment by vigna** [comment]
I just committed a nok branch in which there's no lifetime and it works like a charm. Am I missing something?

**Comment by RagnarGrootKoerkamp** [comment]
Ah nice! It may well be possible to remove it. I was fiddling quite bit with the KeyT trait. I expect the 'k lifetime sn't really needed indeed. Will check out your code later on pc.

**Comment by RagnarGrootKoerkamp** [comment]
Right yes. This is indeed a leftover from fighting the borrow checker for some specific issue that I found solved in another way in the end. Will merge your commit.

**Comment by rob-p** [comment]
But are there cases where a key may need a lifetime reference? Is this something that should be allowed?  If so, should it be preserved somehow in serialization, or omitted?

**Comment by vigna** [comment]
I don't think so because there's the implicit lifetime you get by passing the reference.

BTW, you might consider using a Lender when reading the input, as using an Iterator, say, reading the input from a file, will allocate a lot of strings. Or maybe you have already an in-house solution for that.

**Comment by RagnarGrootKoerkamp** [comment]
I have this `new_from_par_iter` function:
```rust
pub fn new_from_par_iter<'a>(
        n: usize,
        keys: impl ParallelIterator<Item = impl Borrow<Key>> + Clone + 'a,
        params: PtrHashParams,
    )
```

Would that work for you? I'm not quite sure whether each iterator can be converted into a corresponding ParallelIterator that simply doesn't parallellize. This does already take `impl Borrow<Key>` so you can iterate over either `String` or `&[u8]`.

**Comment by vigna** [comment]
> BTW, I didn't notice there was a 128-bit version of xxh3—I ditched SpookyHash in favor of that, with measurable advantages. It seems to be an excellent candidate for 128-bit output. WyHash is currently the fastest (among the good ones) for 64-bit output.

Nah, it does not work—not enough random. Indeed, it fails an SMHasher test. If I try to use xxh3 the probabilistic construction don't work at 100B keys. Bummer.

**Comment by RagnarGrootKoerkamp** [comment]
Oh that's unfortunate. At 1e11 ~ 2^36.5, you only really need 2*36.5=73 bits of entropy, but I suppose the input follows some specific pattern that causes collisions.

- What keys are you running on?
- Is this with PTRHash? If so, I could possibly change the way I extract entropy from the hash. Currently I use the high 64bit word for shard/partition/block, and the low 64bit word for determining the slot, but maybe things are better if I swap the two words.

**Comment by RagnarGrootKoerkamp** [comment]
Ok so here are some timings for UK2005 dataset used in PTHash papers (~40M URLs; ~71byte average length).
```
| Hasher  | Bits | seq. | prefetch |
| fxhash  |   64 | 31.7 |    21.26 |
| murmur2 |   64 | 11.5 |     6.23 |
| murmur3 |  128 | 24.7 |    15.35 |
| wy      |   64 | 19.5 |    11.58 |
| xx      |   64 | 23.2 |    14.26 |
| xx      |  128 | 21.0 |    12.90 |
| city    |   64 | 18.0 |    10.97 |
| city    |  128 | 27.5 |    17.90 |
| metro   |   64 | 22.2 |    15.36 |
| metro   |  128 | 24.3 |    17.71 |
| highway |   64 | 61.1 |    36.23 |
| highway |  128 | 67.0 |    42.51 |
| spooky  |   64 | 74.1 |    62.48 |
| spooky  |  128 | 74.1 |    62.87 |
```
Xx seems to be the fastest 128bit hash, so it's unfortunate that it's not high enough quality. (Note though that I think SSHasher only tests a 64bit variant, not the 128bit variant.)

Murmur2 is otherwise super fast for strings, but I suppose it's quality isn't very high, and similarly murmur3 is the next fastest 128bit hash but it also may not be sufficient?
Otherwise there's City which does pass all SMHasher tests in the 128bit variant.

It's interesting though that at this point a good hash function **triples** the lookup time, so basically the entire problem is reduced to finding _some_ fast injective function from keys into 64 or 128bit hashes.

**Comment by RagnarGrootKoerkamp** [comment]
Anyway, 128bit hashes are basically working now. I just included all of them for now in `lib/hash.rs` so you can try them out if you want.

There isn't much more that I can do -- designing a better hash function is out of scope for me at the moment :)

Last thing may be some more usage examples and API improvements, but those can come later.

**Comment by vigna** [comment]
> Oh that's unfortunate. At 1e11 ~ 2^36.5, you only really need 2*36.5=73 bits of entropy, but I suppose the input follows some specific pattern that causes collisions.
> 
> * What keys are you running on?

Integer [0..s).

> * Is this with PTRHash? 

No no no. Sorry, that was confusing... I switched to xxh3hash in VFunc from sux-rs but there are subtle non-randomness issues that turn out into the random construction not working. It's however a cautionary tale because up 10^9 everything worked fine.

**Comment by vigna** [comment]
Is the code for this test in the repo? I'd like to try my implementation.

**Comment by RagnarGrootKoerkamp** [comment]
Ok for consecutive integers `FxHash` would work but that's cheating :) I'll try `XxHash` as well here.

Yes I'm also confused by the speed differences. My hypothesis is that the crates I'm pulling in are simply not highly optimized. But then again, why implement a specific hash function and not optimize it. It could also have to do with inlining maybe. Possibly they are so large that they are not inlined and hence become slow.

But yes, would be great if you could run locally. `run.rs` now contains `bench_hashers` which does this. (The output is a bit messy though.)
You can add your spooky implementation to `hash.rs` and `bench_hashers` in `run.rs` and you should be good to go.

**Comment by vigna** [comment]
Maybe in a branch? I've checked out master but it's not there.

**Comment by RagnarGrootKoerkamp** [comment]
see `src/bin/run.rs`: https://github.com/RagnarGrootKoerkamp/PTRHash/blob/master/src/bin/run.rs#L172

**Comment by RagnarGrootKoerkamp** [comment]
Allright i'm buliding PTRHash on keys `0..10^11` and it's at 10% now after 10 minutes without collisions so far. 

Code here: https://github.com/RagnarGrootKoerkamp/PTRHash/blob/master/examples/many_keys.rs

**Comment by vigna** [comment]
> see `src/bin/run.rs`: https://github.com/RagnarGrootKoerkamp/PTRHash/blob/master/src/bin/run.rs#L172

How am I supposed to run that from the command line?

**Comment by RagnarGrootKoerkamp** [comment]
`cargo run -r --bin run`
(and `cargo run -r --example many_keys` for the example)

**Comment by vigna** [comment]
> Allright i'm buliding PTRHash on keys `0..10^11` and it's at 10% now after 10 minutes, without collisions so far.
> 
> Code here: https://github.com/RagnarGrootKoerkamp/PTRHash/blob/master/examples/many_keys.rs
> 
> edit: ugh ugh ugh; I forgot to use xxh3. Here we go again :P
> 
> edit 2: oh, indeed xxh3 (128bit) basically fails instantly. Same for murmur2 (64bit), and murmur3 (128bit).

In which sense it fails, and which sense instantly?

**Comment by vigna** [comment]
> cargo run -r --bin run -- -n 1000000000

```
✦ at 17:12:51 ❯ cargo run -r --bin run -- -n 1000000000
    Finished release [optimized + debuginfo] target(s) in 0.01s
     Running `target/release/run -n 1000000000`
error: unexpected argument '-n' found

Usage: run <COMMAND>

For more information, try '--help'.
```

**Comment by RagnarGrootKoerkamp** [comment]
Oh right sorry; do `cargo run -r --bin run -- query -n 1000000000`.

**Comment by RagnarGrootKoerkamp** [comment]
I'm not quite sure how _exactly_ it fails. So first I was using `FxHash` which is simply 64-bit multiplication by a constant. That was working great, probably because keys are distributed over buckets about as smoothly as possible. But actually all other hash functions I try seem to break down, ending up in an 'infinite' (too long) loop while displacing filled buckets. Not quite sure yet why exactly that happens.

Maybe I made a bug for the 128bit hashes, or maybe it's because some bits of the hash are ignored which leads to throwing away too much entropy.

**Comment by vigna** [comment]
> Oh right sorry; do `cargo run -r --bin run -- query -n 1000000000`.

No.

```
✦ at 17:16:40 ❯ cargo run -r --bin run -- query -n 10000000
    Finished release [optimized + debuginfo] target(s) in 0.01s
     Running `target/release/run query -n 10000000`
┌   gen keys:          0.00s
├      clone:          0.00s
├       sort:          0.02s
├ duplicates:          0.00s
generatekeys:          0.04s
        keys:   10000000
      shards:          1
       parts:         39
   slots/prt:     262144
   slots tot:   10223616
 buckets/prt:     101142
 buckets tot:    3944538
 keys/bucket:          2.54
bits/element:          3.42  (pilots 3.16, remap 0.26)
BENCHMARKING loops 30
 ( 1  )  :  2.8
 ( 1  )+r:  3.2
 (64t1)  :  1.91ns
 (64t1)+r:  6.10ns
 (64t2)  :  0.97ns
 (64t2)+r:  3.06ns
 (64t3)  :  0.67ns
 (64t3)+r:  2.05ns
 (64t4)  :  0.50ns
 (64t4)+r:  1.59ns
 (64t5)  :  1.01ns
 (64t5)+r:  2.26ns
 (64t6)  :  0.86ns
 (64t6)+r:  1.80ns
```

**Comment by RagnarGrootKoerkamp** [comment]
That's the output of `benchmark_queries`, not `bench_hashers`. Can you check you're in sync with `master` here?
https://github.com/RagnarGrootKoerkamp/PTRHash/blob/master/src/bin/run.rs#L172
this one looks good to me.

**Comment by vigna** [comment]
> I'm not quite sure how _exactly_ it fails. So first I was using `FxHash` which is simply 64-bit multiplication by a constant. That was working great, probably because keys are distributed over buckets about as smoothly as possible. But actually all other hash functions I try seem to break down, ending up in an 'infinite' (too long) loop while displacing filled buckets. Not quite sure yet why exactly that happens.
> 
> Maybe I made a bug for the 128bit hashes, or maybe it's because some bits of the hash are ignored which leads to throwing away too much entropy.

As a side note, almost all test fail, and they always failed. Are you sure the outcome is correct?
```
---- test::construct stdout ----
┌   gen keys:          0.23s
├      clone:          0.00s
├       sort:          0.00s
├ duplicates:          0.00s
generatekeys:          0.23s
        keys:          2
      shards:          1
       parts:          1
   slots/prt:     262144
   slots tot:     262144
 buckets/prt:         18
 buckets tot:         18
 keys/bucket:          0.11
thread 'test::construct' panicked at src/lib.rs:289:18:
attempt to subtract with overflow
```

**Comment by RagnarGrootKoerkamp** [comment]
oh thanks for the note -- I always run in `-r` and there they (used to, at least) pass but I'll investigate.

Btw, should we switch to a faster chat somewhere?

**Comment by vigna** [comment]
> oh thanks for the note -- I always run in `-r` and there they (used to, at least) pass but I'll investigate.
> 
> Btw, should we switch to a faster chat somewhere?

I'm an idiot, I'm my fork. 🤦🏻‍♂️

**Comment by RagnarGrootKoerkamp** [comment]
Fixed the tests. Was mostly a `usize` constant that was actually negative, and some rounding issues.

**Comment by vigna** [comment]
> oh thanks for the note -- I always run in `-r` and there they (used to, at least) pass but I'll investigate.
> 
> Btw, should we switch to a faster chat somewhere?

Mmmmhhh... Discord? I have installed it, I should be able to use it.

**Comment by RagnarGrootKoerkamp** [comment]
Ok so it seems one reason 128bit hashes were failing was because running with `c=9` (ie `9*n/lg(n)` buckets/key) is not sufficient for such large `n`. Running with `c=10` found no collisions in the first 1% of the shards, and in fact simple Murmur3 also doesn't run into collisions yet this far.

(Building the full thing takes too long for my train ride :D )

I don't really use discord myself. Twitter? Matrix?
